File: essets/js/afne.js
class AFNe {
    constructor(estados, alfabeto, transicoes, estado_inicial, estados_aceitacao) {
        this.estados = estados;
        this.alfabeto = alfabeto;
        this.transicoes = transicoes;
        this.estado_inicial = estado_inicial;
        this.estados_aceitacao = estados_aceitacao;
    }

    epsilon_fecho(estado) {
        let fecho = new Set([estado]);
        let stack = [estado];
        while (stack.length > 0) {
            let current = stack.pop();
            for (let next_state of this.transicoes.get(`${current},`) || new Set()) {
                if (!fecho.has(next_state)) {
                    fecho.add(next_state);
                    stack.push(next_state);
                }
            }
        }
        return fecho;
    }

    aceita(palavra) {
        let estados_atuais = this.epsilon_fecho(this.estado_inicial);
        for (let simbolo of palavra) {
            let novos_estados = new Set();
            for (let estado of estados_atuais) {
                let transicoes_estado = this.transicoes.get(`${estado},${simbolo}`) || new Set();
                for (let next_state of transicoes_estado) {
                    let fecho_next = this.epsilon_fecho(next_state);
                    fecho_next.forEach(s => novos_estados.add(s));
                }
            }
            estados_atuais = novos_estados;
        }
        return Array.from(estados_atuais).some(estado => this.estados_aceitacao.has(estado));
    }
}

// Definindo um AFNe para reconhecer a linguagem {0, 1}* onde o último símbolo é sempre 1.
const estados = new Set(['q0', 'q1', 'q2']);
const alfabeto = new Set(['0', '1']);
const transicoes = new Map([
    ['q0,', new Set(['q1'])],
    ['q0,0', new Set(['q0'])],
    ['q0,1', new Set(['q0'])],
    ['q1,1', new Set(['q2'])]
]);
const estado_inicial = 'q0';
const estados_aceitacao = new Set(['q2']);

const afne = new AFNe(estados, alfabeto, transicoes, estado_inicial, estados_aceitacao);

// Testando algumas palavras
console.log(afne.aceita('101'));   // Saída: true
console.log(afne.aceita('010'));   // Saída: false
console.log(afne.aceita('111'));   // Saída: true

File: AFNe.py
class AFNe:
    def __init__(self, estados, alfabeto, transicoes, estado_inicial, estados_aceitacao):
        self.estados = estados
        self.alfabeto = alfabeto
        self.transicoes = transicoes
        self.estado_inicial = estado_inicial
        self.estados_aceitacao = estados_aceitacao

    def epsilon_fecho(self, estado):
        fecho = {estado}
        stack = [estado]
        while stack:
            current = stack.pop()
            for next_state in self.transicoes.get((current, ''), set()):
                if next_state not in fecho:
                    fecho.add(next_state)
                    stack.append(next_state)
        return fecho

    def aceita(self, palavra):
        estados_atuais = self.epsilon_fecho(self.estado_inicial)
        for simbolo in palavra:
            novos_estados = set()
            for estado in estados_atuais:
                transicoes_estado = self.transicoes.get((estado, simbolo), set())
                for next_state in transicoes_estado:
                    fecho_next = self.epsilon_fecho(next_state)
                    novos_estados.update(fecho_next)
            estados_atuais = novos_estados
        return any(estado in self.estados_aceitacao for estado in estados_atuais)


# Exemplo de utilização:

# Definindo um AFNe para reconhecer a linguagem {0, 1}* onde o último símbolo é sempre 1.
estados = {'q0', 'q1', 'q2'}
alfabeto = {'0', '1'}
transicoes = {
    ('q0', ''): {'q1'},
    ('q0', '0'): {'q0'},
    ('q0', '1'): {'q0'},
    ('q1', '1'): {'q2'}
}
estado_inicial = 'q0'
estados_aceitacao = {'q2'}

afne = AFNe(estados, alfabeto, transicoes, estado_inicial, estados_aceitacao)

# Testando algumas palavras
print(afne.aceita('101'))   # Saída: True
print(afne.aceita('0101'))  # Saída: False
print(afne.aceita('111'))   # Saída: True
